Sources/Socket.swift

```diff
@@ -3226,38 +3268,44 @@ public class Socket: SocketReader, SocketWriter {
 		self.readBuffer.initialize(to: 0x0)
 		memset(self.readBuffer, 0x0, self.readBufferSize)
 		var address: Address? = nil
+
+		var recvFlags: Int32 = 0
+		if (self.readStorage.length > 0) {
+			recvFlags |= Int32(MSG_DONTWAIT)
+		}
 		
 		let addr = sockaddr_storage()
 		var length = socklen_t(MemoryLayout<sockaddr_storage>.size)
 		var addrPtr = addr.asAddr()
-		
+
 		// Read all the available data...
 		#if os(Linux)
-			let count = Glibc.recvfrom(self.socketfd, self.readBuffer, self.readBufferSize, 0, &addrPtr, &length)
+			let count = Glibc.recvfrom(self.socketfd, self.readBuffer, self.readBufferSize, recvFlags, &addrPtr, &length)
 		#else
-			let count = Darwin.recvfrom(self.socketfd, self.readBuffer, self.readBufferSize, 0, &addrPtr, &length)
+			let count = Darwin.recvfrom(self.socketfd, self.readBuffer, self.readBufferSize, recvFlags, &addrPtr, &length)
 		#endif
-		
+
 		// Check for error...
 		if count < 0 {
-			
+
 			// - Could be an error, but if errno is EAGAIN or EWOULDBLOCK (if a non-blocking socket),
 			//		it means there was NO data to read...
 			if errno == EAGAIN || errno == EWOULDBLOCK {
-				
-				return (0, address)
+
+				// FIXME: If we reach this point because data is available in the internal buffer, we will be *unable* to associate it with an address...
+				return (self.readStorage.length, nil)
```

Comment:
Yeah I agree that's probably the best behavior.

Thanks for your help and guidance! I started writing my own socket library for Swift before I realized that there was likely already a solution out there, so finding BlueSocket was a big relief. If I'm struck by anything over the next couple weeks I may send in a PR, but no pressure to check in while you're on break.

Have a great vacation! 😃